### ISSUE TITLE
Bug: TypeScript AST node type `PropertySignature` as a `no-multi-spaces` exception does not resolve conflict with `key-spacing`

### ISSUE THREAD

**Issue by danwithabox** [eslint-stylistic/eslint-stylistic#245]
### Validations

- [X] Follow our [Code of Conduct](https://github.com/eslint-stylistic/eslint-stylistic/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guide](https://eslint.style/contribute/guide).
- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.

### Describe the bug

Given this TS type literal:
```ts
type TMyObj = {
    id:     number,
    code?:  string,
};
```
and the rules:

```ts
"@stylistic/no-multi-spaces": ["error"],
"@stylistic/key-spacing": ["error", {
    "singleLine": {
        "beforeColon": false,
        "afterColon":  true,
    },
    "multiLine": {
        "beforeColon": false,
        "afterColon":  true,
    },
    "align": {
        "beforeColon": false,
        "afterColon":  true,
        "on":          "value",
    },
}],
```
the following is given for `id:     number,`:

> Multiple spaces found before 'number'. eslint[@stylistic/no-multi-spaces](https://eslint.style/rules/js/no-multi-spaces)

Upon resolving that, we get into an inconsitency loop with:

> Missing space before value for key 'id'. eslint[@stylistic/key-spacing](https://eslint.style/rules/ts/key-spacing)

The two rules are in conflict, the documentation suggests using the `exceptions` option of `no-multi-spaces` to resolve such an conflict by ignoring given AST node types.

I checked the node type, it seems to be named `PropertySignature`: [TypeScript AST Viewer](https://ts-ast-viewer.com/#code/C4TwDgpgBAKgsiA8gIwFZQLxQN4CgoFQCWAJgFyEEB2ArgLbIQBOANPoQMYD2JEA-BSgBnYEyJUA5mwC+AbiA)

Updating the rule as per documantion to:
```ts
"@stylistic/no-multi-spaces": ["error", { exceptions: { "Property": true, "PropertySignature": true, } }],
```
does not resolve the issue.
I tried other, adjacent node types as well (`TypeLiteral`, `Identifier`) just in case, to no avail.

I do have `@typescript-eslint/parser` configured as the parser.

Does `exceptions` not allow the use of TypeScript AST nodes?

Repro on stackblitz attached - interestingly, running eslint fix there does nothing even though the issues are reported by it, but manually removing and adding back the spaces does show the inconsistency.

### Reproduction

https://stackblitz.com/edit/eslint-stylistic-eslint-stylistic-ueip7q?file=playground%2Feslint.config.js

### Contributes

- [ ] I am willing to submit a PR to fix this issue
- [ ] I am willing to submit a PR with failing tests

**Comment by tetarchus** [comment]
It seems that using `@typescript-eslint/parser` as a parser, changes the Node type of some of Nodes, including for `interface`s and `type`s.

If you use the [AST Explorer](https://astexplorer.net/) and switch the parser to `@typescript-eslint/parser` and re-inspect the `PropertySignature` node, it is called `TSPropertySignature`. 

Adding `TSPropertySignature` in the `exceptions` key resolved the issue for me.

**Comment by danwithabox** [comment]
Great input, thank you, wouldn't have guessed that `@typescript-eslint/parser` prefixes node types with `TS`!

Interestingly, swapping `PropertySignature` to `TSPropertySignature` did **not** fix it for me.

Out of curiosity, I added `TSTypeAnnotation`, and that did "fix" it in the sense that the inconsistency no longer occured, but of course this breaks type annotations everywhere else, e.g. extra spaces in variable declarations are now permitted.

For sanity, I checked with `TypeAnnotation`, and that does not work as now expected due to `@typescript-eslint/parser`.
So the `TS`-prefixed node type is correct, and is picked up, _except_ for `TSPropertySignature`?

I admit I am rather confused at this point.
Triple-checked everything, can't yet see the reason for this quirk on my end.

> Adding `TSPropertySignature` in the `exceptions` key resolved the issue for me.

Could you please provide a reproduction of the fix?

**Comment by tetarchus** [comment]
It fixed it in terms of type declaration files, but unfortunately as soon as I started going out to other areas of the code base (after posting the original comment) - namely a function with an explicit return type - the spacing started messing up because the `@stylistic/key-spacing` rule doesn't _also_ have the same level of customisation.

I'm still playing around with eslint configs at the moment to try and get a nice replacement for prettier that allows for custom formatting, which will likely involve some custom rules (I don't have experience with writing eslint rules at the moment), so once I've done that I have a bit more knowledge around the plugins I may try submitting a PR for this.

**Comment by jgburet** [comment]
Hey, I'm having the same issue.

In this case, formatting won't do anything (expected), yet it's complaining.
![Screenshot 2024-03-30 at 12 12 58](https://github.com/eslint-stylistic/eslint-stylistic/assets/1304987/67322208-d832-4122-bfbf-c1a51565be03)

Now, if I am to miss-align the keys
![Screenshot 2024-03-30 at 12 22 23](https://github.com/eslint-stylistic/eslint-stylistic/assets/1304987/6e73b733-7e91-431b-aee7-d6a229f7cb90)
formatting will change the code to
![Screenshot 2024-03-30 at 12 23 16](https://github.com/eslint-stylistic/eslint-stylistic/assets/1304987/d98af888-a98c-4321-95c4-e95d1a377f98)
which isn't correct.

This happens both when too few or too many spaces are present before the colon.

```yml
  "@stylistic/no-multi-spaces":
    - warn
    - exceptions:
        Property: true
        TSProperty: true
  "@stylistic/key-spacing":
    - warn
    - align: "colon"
```

**Comment by c-jaye** [comment]
Thanks for the pointers guys, in my case simply adding `{ TSTypeAnnotation: true }` to exceptions fixed all problems for me with this config:

```
"@stylistic/key-spacing": [
   "error",
   {
      singleLine: {
         beforeColon: false,
         afterColon:  true,
         mode:        "strict",
      },
      multiLine: {
         beforeColon: false,
         afterColon:  true,
         mode:        "strict",
      },
      align: {
         beforeColon: false,
         afterColon:  true,
         mode:        "strict",
         on:          "value",
      },
   },
],
"@stylistic/no-multi-spaces": [
   "error",
   {
      ignoreEOLComments: true,
      exceptions:        {
         Property:           true,
         BinaryExpression:   false,
         VariableDeclarator: true,
         ImportDeclaration:  true,
         TSTypeAnnotation:   true,
      },
      includeTabs: true,
   },
],
```

Before:

![image](https://github.com/eslint-community/eslint-stylistic/assets/69225302/4762d764-afac-4189-8630-8ebf16745742)
![image](https://github.com/eslint-community/eslint-stylistic/assets/69225302/c197e0be-ccd3-4fb0-a790-7912f5bfd305)

After:

![image](https://github.com/eslint-community/eslint-stylistic/assets/69225302/8fdd2d4a-5e87-45bc-b125-e6506557b125)

Haven't found anywhere it doesn't work in my case (yet)

**Comment by jgburet** [comment]
Your solution only works when defining a type, not when actually building an object

**Comment by c-jaye** [comment]
Pretty sure your case is a different problem

![image](https://github.com/eslint-community/eslint-stylistic/assets/69225302/c4d2690e-9c39-46cb-bd5c-7a5d1d63b760)

**Comment by jgburet** [comment]
Yup. I've forked the repo and I was actually trying to understand where there would be a loop hole of something.
Turns out, `TSProperty: true` is actually doing its thing properly when building an object.
In fact, the issue stands with the class decorator: ASTExplorer does not yield anything for it!

**Comment by LastDragon-ru** [comment]
Seems this will fix typescript class decorators... 

```
'@stylistic/no-multi-spaces': [
    'error',
    {
        exceptions: {
            Program: true, // !!!!!
        },
    },
],
```

**Comment by jgburet** [comment]
It does, thanks a lot @LastDragon-ru !

**Comment by AvremelM** [comment]
A possible fix here could be something like the `ignoredNodes` option that the `indent` rule has. 
That could allow `key-spacing`, `no-multi-spaces`, and/or `type-annotation-spacing` rules to specify scenarios that they ignore to avoid conflicts.

Honestly, I don't know why that isn't a more common eslint option anyway. Would definitely help resolve many rule conflicts.
